crates/wasm-compose/src/encoding.rs

```diff
@@ -830,7 +882,9 @@ impl<'a> TypeEncoder<'a> {
                 .cur
                 .type_defs
                 .insert(key, state.cur.encodable.type_count());
-            assert!(prev.is_none());
+            // Resource aliases will already have been added to the table, but
+            // other types should not have been:
+            assert!(prev.is_none() || matches!(id, ComponentAnyTypeId::Resource(_)));
```

Comment:
Could this assert that `prev == Some(what_was_inserted)`? 

I also vaguely remember talking about this and somewhat concluding that this is the assertion to have. That being said reading this again I can't immediately determine why it's ok to replace a previous value in the map for resources, so if it's intended could you expand the comment here with why that's ok?

Reply:
This had to do with the `created` vs. `referenced` fields in `wasmparser::types::ComponentEntityType::Type`, which I'll admit I still don't understand.  If you look at `TypeEncoder::component_entity_type` we used `created` for resource types and `referenced` for all other types.  That's the only way I could get it working, but it had the side effect of triggering this assertion.  I'm happy to expand the comment, but I'm afraid I don't understand the situation well enough to explain it.

I'll add the assertion clause you suggested, at least.